### PR TITLE
Improve support to the dictionary returned by ParseQuery

### PR DESCRIPTION
**Pull request by FranklinWhale**
Summary of the changes (Less than 80 chars)
 - Make `AddQueryString(string uri, IEnumerable<KeyValuePair<string, string>> queryString)` `public`
 - Add `AddQueryString(string uri, IEnumerable<KeyValuePair<string, StringValues>> queryString)`
 - Update the summary of `AddQueryString(string uri, IDictionary<string, string> queryString)`
 - Add `QueryBuilder(IEnumerable<KeyValuePair<string, StringValues>> parameters)`

Addresses #7945

### CODE REVIEW

**Comment by FranklinWhale** [comment]
Should `AddQueryString(string uri, IDictionary<string, string> queryString)` be changed to an expression body definition?

i.e.

```C#
public static string AddQueryString(string uri, IDictionary<string, string> queryString)
    => AddQueryString(uri, (IEnumerable<KeyValuePair<string, string>>)queryString);
```

**Review by Tratcher — COMMENTED** [review]
Needs at least one test calling each of the new APIs.

**Comment by Tratcher** [comment]
Reference assemblies need updating.
https://github.com/dotnet/aspnetcore/blob/master/docs/ReferenceAssemblies.md

File: src/Http/Http.Extensions/src/QueryBuilder.cs [hunk 1]
```diff
@@ -23,6 +25,12 @@ public QueryBuilder(IEnumerable<KeyValuePair<string, string>> parameters)
             _params = new List<KeyValuePair<string, string>>(parameters);
         }
 
+        public QueryBuilder(IEnumerable<KeyValuePair<string, StringValues>> parameters)
+            : this(parameters.SelectMany(kvp => kvp.Value, (kvp, v) => KeyValuePair.Create(kvp.Key, v.ToString())))
```
Comment (davidfowl) [inline src/Http/Http.Extensions/src/QueryBuilder.cs]: Looks inefficient. What's the scenario for this overload?

Comment (FranklinWhale) [reply src/Http/Http.Extensions/src/QueryBuilder.cs]: To support the return by [`QueryHelpers.ParseQuery`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.webutilities.queryhelpers.parsequery?view=aspnetcore-3.1)

**Comment by FranklinWhale** [comment]
@Tratcher I think this PR is ready for review again. Thanks!

File: src/Http/Http.Extensions/src/QueryBuilder.cs (hunk 1 above)
Comment (campersau) [inline src/Http/Http.Extensions/src/QueryBuilder.cs]: I think `v` is already a string, so calling `ToString()` on it isn't necessary.
```suggestion
            : this(parameters.SelectMany(kvp => kvp.Value, (kvp, v) => KeyValuePair.Create(kvp.Key, v)))
```

File: src/Http/WebUtilities/test/QueryHelpersTests.cs [hunk 2]
```diff
@@ -119,5 +120,37 @@ public void AddQueryStringWithDictionary(string uri, string expectedUri)
...
+            "http://contoso.com/someaction?name?something",
+            "http://contoso.com/someaction?name?something&param1%5B%5D=value1&param1%5B%5D=&param1%5B%5D=value3&param2%5B%5D=")]
+        [InlineData(
+            "http://contoso.com/someaction#name#something",
+            "http://contoso.com/someaction?param1%5B%5D=value1&param1%5B%5D=&param1%5B%5D=value3&param2%5B%5D=#name#something")]
+        public void AddQueryStringWithEnumerableOfKeysAndStringValues(string uri, string expectedUri)
+        {
+            var queryStrings = new Dictionary<string, StringValues>()
+                        {
+                            { "param1[]", new StringValues(new [] { "value1", string.Empty, "value3" }) },
```
Comment (Tratcher) [inline src/Http/WebUtilities/test/QueryHelpersTests.cs]: The brackets aren't needed here, they make the test data hard to read.
```suggestion
                            { "param1", new StringValues(new [] { "value1", string.Empty, "value3" }) },
```

File: src/Http/WebUtilities/test/QueryHelpersTests.cs [hunk 3]
```diff
@@ -119,5 +120,37 @@ public void AddQueryStringWithDictionary(string uri, string expectedUri)
...
+            "http://contoso.com/someaction?q=test&param1%5B%5D=value1&param1%5B%5D=&param1%5B%5D=value3&param2%5B%5D=#anchor?value")]
+        [InlineData(
+            "http://contoso.com/someaction#anchor?stuff",
+            "http://contoso.com/someaction?param1%5B%5D=value1&param1%5B%5D=&param1%5B%5D=value3&param2%5B%5D=#anchor?stuff")]
+        [InlineData(
+            "http://contoso.com/someaction?name?something",
+            "http://contoso.com/someaction?name?something&param1%5B%5D=value1&param1%5B%5D=&param1%5B%5D=value3&param2%5B%5D=")]
+        [InlineData(
+            "http://contoso.com/someaction#name#something",
+            "http://contoso.com/someaction?param1%5B%5D=value1&param1%5B%5D=&param1%5B%5D=value3&param2%5B%5D=#name#something")]
```
Comment (Tratcher) [inline src/Http/WebUtilities/test/QueryHelpersTests.cs]:
```suggestion
            "http://contoso.com/someaction?param1=value1&param1=&param1=value3&param2=#name#something")]
```

**Review by Tratcher — COMMENTED** [review]
One piece of test feedback and I think this is good to go. Thanks.

**Comment by FranklinWhale** [comment]
@Tratcher I have removed the brackets in the test cases. Unfortunately, getting this PR to pass all CI tests seems very difficult...

**Review by Tratcher — APPROVED** [review]

**Comment by Tratcher** [comment]
Thanks

**Pull request merged**